engine/buff/buff.cpp

```diff
@@ -829,6 +835,80 @@ buff_t* buff_t::set_duration_multiplier( double multiplier )
   return this;
 }
 
+// Applies a modifier from -99 to infinity that controls how fast the buff functions.
+// Values less than -99 will be rounded to -99, which seems to match in-game behavior where
+// auras with a time modifier effect of -100 actually only apply a 100x slowdown, and not a total pause.
+// It's intended to modify real time duration and tickrate, without affecting the apparent duration
+// as used for things like pandemic refresh behavior.
+// Currently, this only modifies the duration of the buff, moving its expiration closer or
+// further out. The "apparent" duration is lost (but recoverable), since remains() returns the
+// real-time duration. The tickrate is un-adjusted, since it is currently based on the real-time duration.
+// None of these limitations particularly matter for current usecases, but if you're looking at using this, be
+// aware there may be more work required to support your usecase.
+buff_t* buff_t::apply_time_modifier_duration( double modifier )
+{
+  if ( ignore_mod_time_modifier )
+    return this;
+  modifier = std::max( modifier, -99.0 ); // Limit slow down to 100x slower
+
+  auto mul = 100.0 / ( 100 + modifier );
+
+  base_time_modifier_duration_multiplier = base_time_modifier_duration_multiplier * mul;
+
+  return this;
+}
```

Comment:
This is a fairly large departure from how multipliers are handled in simc. If the underlying thing you're changing is a multiplier, it should be accessed as a multiplier, not a modifier. I realize that you can't just use `1 + time_mod.percent()` (since you need a reciprocal of that), but this is more of a question of whether to store `mult` or `1 / mult` (see haste%).

Reply:
> This is a fairly large departure from how multipliers are handled in simc

> Tracking the modifier rather than multiplier is very unusual as far as simc goes.

Surely either the modifier from the effect or the multiplier used in the calculations as a result is "correct" to track, and not some derived 3rd thing.

This absolutely is not a datum like other multipliers used in simc, from what I can tell, because the `%`-equivalent of the modifier isn't a straight multiplier on durations, what with how values less than zero are handled. If you'd like, I could add a

```c++
  double time_multiplier() const
  { return ( 100.0 / ( 100.0 + std::max( _base_value, -99.0 ) ) ); }
```

to the `spell_data` type itself, and push the multiplier calculation there, if doing it on-use like this seems somehow lacking...

Reply:
> This absolutely is not a datum like other multipliers used in simc, from what I can tell, because the %-equivalent of the modifier isn't a straight multiplier on durations

Haste and cooldown recharge effects behave the same way. If you have a +50% cooldown recharge rate, you need to transform it into 1/1.5 to use as a multiplier. Same with haste.

> especially with the apparent hard cap at -99% per-effect

Same cap exists for damage modifiers (if you have -100% damage taken, the game caps it at -99%). The main point here is that I'm not convinced we need to depart from the standard multiplier "interface." Why not just do this at the point of use and only make sure the user doesn't try to make the multiplier 0?

Reply:
> Why not just do this at the point of use and only make sure the user doesn't try to make the multiplier 0?

While individual effects seem to be capped at -99% (based on how the UHUH buffs (which have a -100% effect value but only apply a 100x slowdown, not an infinite one) actually scale time and are reported by other APIs), the aggregate multiplier is able to be much higher - a UHUH-paused buff in an elisande boss slow field is observed at a 10000x slowdown. So yeah, the multiplier obviously has to be positive and nonzero, but the modifier itself also has to be capped when you create the multiplier from it to emulate how stacking appears to work.

Reply:
> Haste and cooldown recharge effects behave the same way.

Ignoring the seemingly unique per-effect capping behavior:

* `haste` is stored as the raw multiplier, and every application of an effect does the `haste *= 1.0 / (1.0 + effect.percent())` transform. Which seems like a footgun, but alright (maybe we should build that transform into the spell data type, so you can just do `effect.multiplier()`!).
 
* Cooldown recharge, meanwhile, _is_ stored as a reciprocal multiplier. So every effect application does `base_recharge_multiplier *= 1 + effect.percent()`. Except for `architects_ingenuity_t` which stores the multiplier instead.

The two seem to have picked different options here - looking to them for guidance, both the multiplier _and_ the reciprocal multiplier seem fine. But there's definitely a preference for the multiplier and not the reciprocal - `haste` is used far more than cooldown recharge, and cooldown recharge isn't fully consistent anyway.

The only difference currently here from how `haste` behaves would seem to be that the `apply_modifier` spell data loading function does the transform from modifier to multiplier (whichever it is), rather than the use-site? Which isn't actually uncommon, since the local function inside spell data loading, `apply_percent_effect_modifier` also does that? (And tbh, the apply function I have here *could* be a local function inside spell data loading - maybe it should be.)